src/averageAges.js

```diff
@@ -55,10 +82,34 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const childrensWithMoms = people
+    .filter(({ mother }) => mother !== null);
+
+  const menWithMoms = people
+    .filter(({ mother, sex }) => mother !== null && sex === SEX_MALE);
+
+  const childrens = onlyWithSon ? menWithMoms : childrensWithMoms;
+
+  const ageDifferences = childrens.map(({ mother: motherName, born }) => {
+    const mother = people.find(({ name }) => name === motherName);
+
+    return mother ? born - mother.born : null;
+  });
+
+  const validAgeDifferences = ageDifferences.filter(el => el !== null);
+  const averageAgeDifference = calculateAverageAge(validAgeDifferences);
+
+  return averageAgeDifference;
+};
+
+function calculateAgeDifferences(people) {
+  return people.map(({ died, born }) => died - born);
+}
+
+function calculateAverageAge(ageDifferences) {
+  const totalAgeDifference = ageDifferences.reduce((acc, value) => acc + value);
+
+  return totalAgeDifference / ageDifferences.length;
```

Comment:
These functions can be combined into one, so as not to make an unnecessary pass through the array. To do this, you can transfer the logic of calculations from `map` to the middle of `reduce`.
Then this new feature can be used for calculations men and women average age.